### PR TITLE
 Fixes issue #175 retrying on Scan operation on error

### PR DESCRIPTION
**Pull request by BharathMG**
Fixes scan never getting started when retrying on failed scan operation

### CODE REVIEW

**Comment by BharathMG** [comment]
Sure @uKL . I will add it.

**Comment by uKL** [comment]
Thanks a lot. Great job addressing this bug! :) 🥇

**Comment by BharathMG** [comment]
@uKL My pleasure! Thanks for this wonderful library :) 

Added a couple of tests. Let me know if I should change anything.

**Comment by BharathMG** [comment]
Got it. I will make the necessary changes.

**Comment by BharathMG** [comment]
@dariuszseweryn Will this approach be fine?

      
        given:
        TestSubscriber testSubscriber = new TestSubscriber<>()
        bleAdapterWrapperSpy.hasBluetoothAdapter() >> true
        bleAdapterWrapperSpy.isBluetoothEnabled() >> true

        when:
        def scanObservable = objectUnderTest.scanBleDevices(null)

        then:
        0 * bleAdapterWrapperSpy.isBluetoothEnabled()

        when:
        scanObservable.subscribe(testSubscriber)

        then:
        1 * bleAdapterWrapperSpy.startLeScan(_) >> true

File: rxandroidble/src/test/groovy/com/polidea/rxandroidble/RxBleClientTest.groovy [hunk 1]
```diff
@@ -79,6 +79,21 @@ class RxBleClientTest extends Specification {
         1 * bleAdapterWrapperSpy.startLeScan(_) >> true
     }
 
+    def "should start BLE scan if bluetooth is enabled before subscribing to the scan observable"() {
+        given:
+        TestSubscriber testSubscriber = new TestSubscriber<>()
+        bleAdapterWrapperSpy.hasBluetoothAdapter() >> true
+        bleAdapterWrapperSpy.isBluetoothEnabled() >> false
+
+        when:
```
Comment (BharathMG) [inline rxandroidble/src/test/groovy/com/polidea/rxandroidble/RxBleClientTest.groovy]: @dariuszseweryn Will this approach be fine?

      
        given:
        TestSubscriber testSubscriber = new TestSubscriber<>()
        bleAdapterWrapperSpy.hasBluetoothAdapter() >> true
        bleAdapterWrapperSpy.isBluetoothEnabled() >> true

        when:
        def scanObservable = objectUnderTest.scanBleDevices(null)

        then:
        0 * bleAdapterWrapperSpy.isBluetoothEnabled()

        when:
        scanObservable.subscribe(testSubscriber)

        then:
        1 * bleAdapterWrapperSpy.startLeScan(_) >> true

Comment (BharathMG) [inline rxandroidble/src/test/groovy/com/polidea/rxandroidble/RxBleClientTest.groovy]: @dariuszseweryn Will this approach be fine? 

      
        given:
        TestSubscriber testSubscriber = new TestSubscriber<>()
        bleAdapterWrapperSpy.hasBluetoothAdapter() >> true
        bleAdapterWrapperSpy.isBluetoothEnabled() >> true

        when:
        def scanObservable = objectUnderTest.scanBleDevices(null)

        then:
        0 * bleAdapterWrapperSpy.isBluetoothEnabled()

        when:
        scanObservable.subscribe(testSubscriber)

        then:
        1 * bleAdapterWrapperSpy.startLeScan(_) >> true

Comment (dariuszseweryn) [inline rxandroidble/src/test/groovy/com/polidea/rxandroidble/RxBleClientTest.groovy]: Yes

**Comment by BharathMG** [comment]
Kindly check and let me know if anything is missing.

File: rxandroidble/src/test/groovy/com/polidea/rxandroidble/RxBleClientTest.groovy [hunk 2]
```diff
@@ -79,6 +79,49 @@ class RxBleClientTest extends Specification {
         1 * bleAdapterWrapperSpy.startLeScan(_) >> true
     }
 
+    def "should start BLE scan if bluetooth is enabled before subscribing to the scan observable"() {
```
Comment (dariuszseweryn) [inline rxandroidble/src/test/groovy/com/polidea/rxandroidble/RxBleClientTest.groovy]: This description does not seem to match the test. Here it is checking if `bleAdapterWrapperSpy.isBluetoothEnabled()` was not called in line 92

File: rxandroidble/src/test/groovy/com/polidea/rxandroidble/RxBleClientTest.groovy [hunk 3]
```diff
@@ -79,6 +79,49 @@ class RxBleClientTest extends Specification {
...
+        0 * bleAdapterWrapperSpy.isBluetoothEnabled()
+
+        when:
+        scanObservable.subscribe(testSubscriber)
+
+        then:
+        1 * bleAdapterWrapperSpy.startLeScan(_) >> true
+    }
+
+    def "should start scan for only subscribers with no exceptions from scan observable"() {
```
Comment (dariuszseweryn) [inline rxandroidble/src/test/groovy/com/polidea/rxandroidble/RxBleClientTest.groovy]: Could you rephrase it to be easier to understand? Maybe something like: `"should check if /condition/ is met at the time of each subscription"`
It could like like this: 
```
        given:
        def firstSubscriber = new TestSubscriber<>()
        def secondSubscriber = new TestSubscriber<>()
        bleAdapterWrapperSpy.hasBluetoothAdapter() >> true
        bleAdapterWrapperSpy.startLeScan(_) >> true
        def scanObservable = objectUnderTest.scanBleDevices(null)

        when:
        scanObservable.subscribe(firstSubscriber)

        then:
        1 * bleAdapterWrapperSpy.isBluetoothEnabled() >> false
        
        and:
        firstSubscriber.assertError {
            BleScanException exception -> exception.reason == BLUETOOTH_DISABLED
        }

        when:
        scanObservable.subscribe(secondSubscriber)

        then:
        1 * bleAdapterWrapperSpy.isBluetoothEnabled(_) >> true
        
        and:
        secondSubscriber.assertNoErrors()
```
Then we would need to make the test exercise for other conditions (`isLocationPermissionOk` / `isLocationProviderOk`). It could be quite easily achieved by using `where:` clause

File: rxandroidble/src/test/groovy/com/polidea/rxandroidble/RxBleClientTest.groovy [hunk 4]
```diff
@@ -216,6 +259,21 @@ class RxBleClientTest extends Specification {
         }
     }
 
+    def "should emit BleScanException if bluetooth was disabled before starting scan"() {
```
Comment (dariuszseweryn) [inline rxandroidble/src/test/groovy/com/polidea/rxandroidble/RxBleClientTest.groovy]: This test is basically redundant to the one above and to the one at line 101

File: rxandroidble/src/test/groovy/com/polidea/rxandroidble/RxBleClientTest.groovy [hunk 5]
```diff
@@ -244,6 +302,21 @@ class RxBleClientTest extends Specification {
         }
     }
 
+    def "should emit BleScanException if location permission was not granted before starting scan"() {
```
Comment (dariuszseweryn) [inline rxandroidble/src/test/groovy/com/polidea/rxandroidble/RxBleClientTest.groovy]: seems also to be redundant

File: rxandroidble/src/test/groovy/com/polidea/rxandroidble/RxBleClientTest.groovy (hunk 3 above)
Comment (BharathMG) [inline rxandroidble/src/test/groovy/com/polidea/rxandroidble/RxBleClientTest.groovy]: Got it. Do you think this would be fine using where? 

        def "should check if all the conditions are met at the time of each subscription"() {
        given:
        def firstSubscriber = new TestSubscriber<>()
        def secondSubscriber = new TestSubscriber<>()
        bleAdapterWrapperSpy.isBluetoothEnabled() >> bluetoothEnabled >> true
        bleAdapterWrapperSpy.hasBluetoothAdapter() >> bluetoothAvailable >> true
        locationServicesStatusMock.isLocationPermissionOk() >>> [locationPermissionsOk, true]
        locationServicesStatusMock.isLocationProviderOk() >>> [locationProviderOk, true]
        def scanObservable = objectUnderTest.scanBleDevices(null)

        when:
        scanObservable.subscribe(firstSubscriber)

        then:
        firstSubscriber.assertError {
            BleScanException exception -> exception.reason == reason
        }

        when:
        scanObservable.subscribe(secondSubscriber)

        then:
        secondSubscriber.assertNoErrors()
        1 * bleAdapterWrapperSpy.startLeScan(_) >> true

        where:
        bluetoothAvailable | bluetoothEnabled | locationPermissionsOk | locationProviderOk | reason
        false              | true             | true                  | true               | BLUETOOTH_NOT_AVAILABLE
        true               | false            | true                  | true               | BLUETOOTH_DISABLED
        true               | true             | false                 | true               | LOCATION_PERMISSION_MISSING
        true               | true             | true                  | false              | LOCATION_SERVICES_DISABLED
    }

Comment (dariuszseweryn) [inline rxandroidble/src/test/groovy/com/polidea/rxandroidble/RxBleClientTest.groovy]: Yup. That should do. We should have also the test that would first pass and then fail (the opposite scenario).

**Comment by dariuszseweryn** [comment]
@BharathMG Thank you for your support :) 👍

**Comment by uKL** [comment]
Dear Contributor,

similar to many open source projects we kindly request to sign our CLA if you'd like to contribute to our project. This license is for your protection as a Contributor as well as the protection of Polidea; it does not change your rights to use your own Contributions for any other purpose.

You can find a link here: https://cla-assistant.io/Polidea/RxAndroidBle
